### PR TITLE
ROMIO: GPFS: fix for Alltoallv potentially using large disps

### PR DESCRIPTION
**Pull request by markalle**
There was a problem with the Alltoallv in ADIOI_GPFS_Calc_others_req().
It was using sendbuf and sdisps[] to reference the various
my_req[x].offsets, but those are all malloced separately and could
thus be further apart than the integers in sdisps[] would allow.
So this PR copies all that data into a new send buf before the
Alltoallv, then copies it back out of a new recv buf.

I don't have a testcase that demonstrates the failure, but I did
at least run a GPFS test that called MPI_File_write_all() and
uses the Alltoallv().

Signed-off-by: Mark Allen <markalle@us.ibm.com>

---------------------------------------------------------------------------------

This was previously in PR https://github.com/pmodels/mpich/pull/4948 where I had too many commits in a single PR, and when I separated them out this one got closed as "this PR has been replaced" but the alltoallv part didn't get replaced, it still needs to go in.

### CODE REVIEW

**Comment by hzhou** [comment]
@markalle I sent you an invite to github team `mpich-other`. It will allow `mpich-warnings` test to run on your PRs.

**Comment by markalle** [comment]
Shoot, I must have not noticed the invite in time and maybe it expired. Try again?

**Comment by hzhou** [comment]
> 
> 
> Shoot, I must have not noticed the invite in time and maybe it expired. Try again?

Sure. But could you first check the mailbox you associated with github and check for email from github? The invite is still pending, so if you find the email like, I think it should still work. If not, let me know.

**Comment by markalle** [comment]
Ah, I see it now, worked, thanks.

**Comment by markalle** [comment]
test:mpich/warnings

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c [hunk 1]
```diff
@@ -641,8 +645,18 @@ void ADIOI_GPFS_Calc_others_req(ADIO_File fd, int count_my_req_procs,
...
+#define HAS_ALLTOALLV_C 0
+#endif
     /* Parameters for MPI_Alltoallv */
-    int *scounts, *sdispls, *rcounts, *rdispls;
+    MPI_Aint *sdispls, *rdispls;
+#if HAS_ALLTOALLV_C
+    MPI_Count *scounts, *rcounts;
+#else
+    int *scounts, *rcounts;
+#endif
```
Comment (hzhou) [inline src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c:659]: Just move the declarations to where it used (Malloc).

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c [hunk 2]
```diff
@@ -641,8 +645,18 @@ void ADIOI_GPFS_Calc_others_req(ADIO_File fd, int count_my_req_procs,
     int i;
     ADIOI_Access *others_req;
 
+#if MPI_VERSION >= 1234 || (MPI_VERSION == 1234 && MPI_SUBVERSION > 1234)
```
Comment (hzhou) [inline src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c:648]: What is `1234`? `>= 4` should work, right?

**Review by hzhou — CHANGES_REQUESTED** [review]
I think this PR is good. Just some minor comments. The PR and commit title need change. The title needs identity what is the code change, not its back story. Also please rebase to main branch and get rid of the merge commits.

**Comment by roblatham00** [comment]
GPFS's ALLTOALL optimization is good but can exhaust memory on some platforms, so I am a little wary (only a little!) of adding more memory allocations. Your approach is fine --  I hope any system that balks at an additional megabyte of memory usasge (100,000 processes * 4 * 2) will use MPI-4!

Hui is suggesting the title be something like `ROMIO: GPFS: fix possibly large displacements in ALLTOALLV` .    In fact the title in your OpenMPI commit is fine https://github.com/open-mpi/ompi/commit/d85cac8f1a114

Thank you for paying attention to ROMIO's GPFS driver and for contributing fixes back.

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c (hunk 1 above)
Comment (markalle) [reply src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c]: Good idea, I moved it

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c (hunk 2 above)
Comment (markalle) [reply src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c]: I had just put that as a placeholder, but I guess MPI-4 is getting close enough to official to say >= 4

**Comment by markalle** [comment]
test:mpich/warnings

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c (hunk 2 above)
Comment (hzhou) [reply src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c]: MPI-4 is fully implemented in the main branch.

**Comment by wkliao** [comment]
FYI. I notice a page from the User Guide of Summit at OLCF, Section
Title: "Parallel I/O crash on GPFS with latest MPI ROMIO". 
Their fix is to use non-blocking MPI calls.
https://docs.olcf.ornl.gov/systems/summit_user_guide.html#known-issues

**Comment by roblatham00** [comment]
It looks like Mark has addressed all of Hui's changes.  I rebased to master and pushed a new version with some compiler warning fixes.

**Review by hzhou — APPROVED** [review]
LGTM

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch3/tcp

**Comment by roblatham00** [comment]
trivial rebase.   tests passed.

**Pull request merged**